Add auto buffer to netplay

Aestek opened the pull request:
Add option to automatically adapt buffer based on last ping values.
Buffer is calculated using a 0.8 median to ignore lag spikes but still
account for ping fluctuations.

CoolOppo commented:
This is something that, if done correctly, will make Netplay much smoother with less involvement.

I'm reviewing the change right now, and there need to be changes made before this could be considered "done correctly". I like a lot of what you did; there are just a few things that need to be addressed.

I'm trying my best to be very detailed about everything because I feel *very* strongly that this feature is crucial to improving Netplay.

lioncash commented:
*[Source/Core/Core/NetPlayServer.h, line 92 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPMAE2-6yvfjvId8G0:-KLPMAE3HquiFK1PlnTr:89729570) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.h#L92)):*
> ```objc
>   void UpdateBuffer();
> 
>   static constexpr const u32 AUTO_BUFFER_PING_WINDOW = 3;
> ```

`const` isn't necessary for both of these.

Also this should likely indicate what the `window` scale actually is (e.g. is it seconds, milliseconds, etc, etc) either in the name or as a constant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964)*
<!-- Sent from Reviewable.io -->

CoolOppo commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/NetPlayServer.cpp, line 764 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPSBE0kQb29ohx1sAH:-KLPSBE0kQb29ohx1sAI:-2143866162) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L764)):*
> ```cpp
> 
>   u32 window = AUTO_BUFFER_PING_WINDOW * static_cast<u32>(m_players.size());
>   if (size < window)
> ```

If `UpdateBuffer()` is only called from `HandlePing()`, you don't need to check this condition because it has already been handled in `HandlePing()` (at line 752).

---

*[Source/Core/Core/NetPlayServer.cpp, line 771 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLP7u8Qhtk9chMORa2M:-KLP7u8RIQAFIrIuVpS6:1761048935) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L771)):*
> ```cpp
> 
>   float divisor = 1 / AUTO_BUFFER_SENSIBILITY;
>   size_t index = static_cast<size_t>(round(size / divisor));
> ```

It's very odd to define `divisor` as the reciprocal of `AUTO_BUFFER_SENSIBILITY` and then divide  `size` by it. That's the same as multiplying the two (basic algebra), so instead of defining another variable, you could simply do:

```cpp
size_t index = static_cast<size_t>(round(size * AUTO_BUFFER_SENSIBILITY)); // Still incorrect, see below!
```

Additionally, you have a bad off-by-one error. You should have `(size - 1)` instead of `size`.

So, this line should look like this:

```cpp
size_t index = static_cast<size_t>(round((size - 1) * AUTO_BUFFER_SENSIBILITY));
```

---

*[Source/Core/Core/NetPlayServer.cpp, line 772 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPLulV4_LygsIJLiA0:-KLPLulV4_LygsIJLiA1:-125182341) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L772)):*
> ```cpp
>   float divisor = 1 / AUTO_BUFFER_SENSIBILITY;
>   size_t index = static_cast<size_t>(round(size / divisor));
>   u32 buffer = last_pings[index] / 8;
> ```

Another issue I have is that you're setting the buffer based on a ping that isn't the maximum ping. The game will stutter if the buffer can not account for the ping, and if one of the players' ping deviates a lot, this means that there is a good chance the buffer will be set below the size that can actually handle a ping that high, and **will** cause a guaranteed stutter.

In other words, don't choose to set the buffer to just account for the third-highest ping received (which is what is happening now). Instead, set the buffer to account for the highest ping received!

Basically, this would be better:

```diff
- u32 window = AUTO_BUFFER_PING_WINDOW * static_cast<u32>(m_players.size());
- if (size < window)
-     return;
-  
- std::deque<u32> last_pings = m_last_pings;
sort(last_pings.begin(), last_pings.end());
- 
- float divisor = 1 / AUTO_BUFFER_SENSIBILITY;
- size_t index = static_cast<size_t>(round(size / divisor));

+ u32 buffer = last_pings[size - 1];
```

This should be done using an [attack and release](https://en.wikipedia.org/wiki/Dynamic_range_compression#Attack_and_release)-esque design, with a slow release. This would ensure that there aren't constantly changes to the buffer, which would be bad, because changing the buffer invisibly makes the input lag fluctuate.

To do so, we need to ~~have some sort of count of how many times the function has run and *wanted* to decrease since the buffer was last **increased**, and only decrease it after it has been a certain number of cycles~~ increase `AUTO_BUFFER_PING_WINDOW` to ensure that it doesn't act too quickly and decrease the pad buffer before we're sure that the ping is stable enough.

---

*[Source/Core/Core/NetPlayServer.h, line 92 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPMAE2-6yvfjvId8G0:-KLPSpQnnVzUqof3cfh7:697630159) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.h#L92)):*
> is it seconds, milliseconds, etc
It's a measure of how many of the last pings from each player are to be considered, so the time that takes just depends on wherever the time between pings is defined.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964)*
<!-- Sent from Reviewable.io -->

CoolOppo commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Core/NetPlayServer.cpp, line 772 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPW0XzcVZFNoBHgJdk:-KLPW0XzcVZFNoBHgJdl:1450487368) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L772)):*
> ```cpp
>   float divisor = 1 / AUTO_BUFFER_SENSIBILITY;
>   size_t index = static_cast<size_t>(round(size / divisor));
>   u32 buffer = last_pings[index] / 8;
> ```

And yes, I know that you originally wanted to use the ".8 median", but I just want to reiterate that doing so means that we're potentially accepting a buffer that is not high enough to account for the latency, which would in turn cause a lag spike.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964)*
<!-- Sent from Reviewable.io -->

Aestek commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Core/NetPlayServer.cpp, line 772 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPW0XzcVZFNoBHgJdk:-KLPs76Az0lk7y1DvQWE:-922638842) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L772)):*
<details><summary><i>Previously, CoolOppo wrote…</i></summary>
> And yes, I know that you originally wanted to use the ".8 median", but I just want to reiterate that doing so means that we're potentially accepting a buffer that is not high enough to account for the latency, which would in turn cause a lag spike.
</details>
Attack and release indeed looks like a good approach to this problem, I'll look into it.

The problem with this feature is that the best solution really depends on the game you are playing. For something like Mario Party you are totally right, we do not want any stutter and a high buffer is not that much of an issue. For games like Smash high buffer is hell and some occasional stutters can be accepted if that means a more stable, low buffer: in my experience it's better to stutter once for a few ms because the ping went to say 30ms to 150ms than see the buffer skyrocket to 12/14  for a potentially much longer time : 3~4s in this case (From a Smash perspective that is). That's why I'm ignoring some of the ping values in the window (btw If you have better wording for ".8 median" I'll take it :p).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964)*
<!-- Sent from Reviewable.io -->

comex commented:
*[Source/Core/Core/NetPlayServer.cpp, line 772 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPW0XzcVZFNoBHgJdk:-KLPwUF5G-5h-Ge3PLJ2:-856170405) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L772)):*
> (btw If you have better wording for ".8 median" I'll take it :p).

80th percentile.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964)*
<!-- Sent from Reviewable.io -->

Aestek commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Core/NetPlayServer.cpp, line 764 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPSBE0kQb29ohx1sAH:-KLPyVKmtAJDhEREwAH-:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L764)):*
<details><summary><i>Previously, CoolOppo wrote…</i></summary>
> If `UpdateBuffer()` is only called from `HandlePing()`, you don't need to check this condition because it has already been handled in `HandlePing()` (at line 752).
</details>
Done.

---

*[Source/Core/Core/NetPlayServer.cpp, line 771 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLP7u8Qhtk9chMORa2M:-KLPyZFqqhrE9wjDvmGS:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L771)):*
<details><summary><i>Previously, CoolOppo wrote…</i></summary>
> It's very odd to define `divisor` as the reciprocal of `AUTO_BUFFER_SENSIBILITY` and then divide  `size` by it. That's the same as multiplying the two (basic algebra), so instead of defining another variable, you could simply do:
> 
> ```cpp
> size_t index = static_cast<size_t>(round(size * AUTO_BUFFER_SENSIBILITY)); // Still incorrect, see below!
> ```
> 
> Additionally, you have a bad off-by-one error. You should have `(size - 1)` instead of `size`.
> 
> So, this line should look like this:
> 
> ```cpp
> size_t index = static_cast<size_t>(round((size - 1) * AUTO_BUFFER_SENSIBILITY));
> ```
> 
</details>
Done.

---

*[Source/Core/Core/NetPlayServer.h, line 92 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPMAE2-6yvfjvId8G0:-KLPyXcslxz_jDlwUZO_:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.h#L92)):*
<details><summary><i>Previously, CoolOppo wrote…</i></summary>
> > is it seconds, milliseconds, etc
> It's a measure of how many of the last pings from each player are to be considered, so the time that takes just depends on wherever the time between pings is defined.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964)*
<!-- Sent from Reviewable.io -->

comex commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/NetPlayServer.cpp, line 772 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964#-KLPxE3kbGNetgLaj1o2:-KLPxE3kbGNetgLaj1o3:-1298910039) ([raw file](https://github.com/dolphin-emu/dolphin/blob/8a77ca0147abb9f13f57770c73e71e50436c0f05/Source/Core/Core/NetPlayServer.cpp#L772)):*
> ```cpp
>   float divisor = 1 / AUTO_BUFFER_SENSIBILITY;
>   size_t index = static_cast<size_t>(round(size / divisor));
>   u32 buffer = last_pings[index] / 8;
> ```

So, as things currently stand, you can't just convert timing to a buffer amount, because the conversion is actually variable: a new input is pushed/popped whenever VideoInterface polls SI (SerialInterface::UpdateDevices -> CSIDevice_GCController::GetData -> ... -> NetPlay_GetInput), and the frequency of that can be set by the game via SI_POLL.

Maybe the packet send rate and game polling rate should be decoupled... but if not, you'll need to figure out a way to correct for rate changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3964)*
<!-- Sent from Reviewable.io -->

Aestek commented:
I reworked this to improve the following:

* @comex poll timings are no longer hardcoded but are taken from HW. This means it handles 50 and 60 fps games. It should also work with weird 120fps hacks but that needs some testing.
* @CoolOppo I reworked the UI a bit to include a slider (see screenshot below) that goes from "Reactive" to "Smooth" that controls the sensibility parameter you pointed out. It goes from .5 to 1 and defaults to .9.
* Buffer calculation now accounts for more than two players for the `client -> server ->client` scenario. The max sum of possible player ping pairs is taken to calculate the buffer. Since the host's ping is virtually 0, for a two players session ping is `p2`. For a three players session ping is `p2 + p3`. For a four players session ping is `max(p2 + p3, p2 + p4, p3 + p4)`. This needs some testing too.

![image](https://cloud.githubusercontent.com/assets/1712219/16472248/01b632fa-3e63-11e6-9367-5ad4079122c5.png)

E2xD commented:
Upon testing this with JMC47, when we sat the buffer to Auto + 100% Smooth, it would amp the buffer up to triple digits/an absurd buffer.

Aestek commented:
@E2xD thx for ur feedback, ill look into that when I get home

Aestek commented:
Pinging @MaJoRoesch because of UI changes (:

Aestek commented:
@E2xD absurd buffer should be fixed :)

E2xD commented:
Unless the PR wasn't updated, the bug is still there for me.

JMC47 commented:
@dolphin-emu-bot rebuild.

JMC47 commented:
@Aestek as I said in IRC, I don't think this actually works.  In Mario Superstar Baseball, playing with a friend on a weaker computer, autobuffer was correct in sellecting 2 buffer for 16ms of latency.  Unfortunately, latency is more than just buffer, so we were constantly lagging due to his CPU not being able to handle it at 2 buffer.  Raising it to 5 buffer made it full speed; but I had to turn on manual for that.

Also the set buffer messages were doubled.

CoolOppo commented:
I don't think we should give up though at all. This is a great feature.

Jofzar commented:
When i was playing with a friend at 14 buffer, its would switch between 10 and 2 buffer and then when ping was higher it was just not reacting and dynamically changing, maybe I missed something.

GonzaloAndrade commented:
Keep the good job guys!

JMC47 commented:
Just a reminder: the github thread is for testing + feedback.  If you don't have anything important for this Pull Request, please comment elsewhere (forums, perhaps?)

MaJoRoesch commented:
@Aestek Can you give me fresh before and afters please?

degasus commented on Source/Core/Core/NetPlayServer.cpp:
```diff
@@ -733,6 +756,57 @@ bool NetPlayServer::StartGame()
+  u32 poll_rate = SerialInterface::GetPollXFrame() > 0 ? SerialInterface::GetPollXFrame() :
+                                                         DEFAULT_PAD_POLL_RATE;
+
+  u32 refresh_rate = VideoInterface::GetTargetRefreshRate() > 0 ?
```
Why do you depend on VI scanout configuration? I'd suggest to not use any emulation state here but maybe GCpad configuration.

CoolOppo commented:
So what's going on with this at the moment? What hurdles do we need to get past before this is ready for production?

ZephyrSurfer commented:
Please rebase.
